Add ND platform to abstraction

rtolos opened the pull request.

tahigash commented on src/rest/connector/libs/nd/implementation.py:
```diff
@@ -0,0 +1,310 @@
+
+        devices:
+            nd1:
+                platform: nd
```
I believe should be `os`. did you test?
```suggestion
                os: nd
```

rtolos replied:
No, it's platform - the abstraction layer works on 'os' or 'platform' - 'os' is fine in most cases, but for Nexus Dashboard nodes (basically, linux-based virtual machines) 'os' should be linux, so platform is used for the abstraction.
Tested, it works as intended

tahigash replied:
it's fine. iosxe is also based on linux. but os is `iosxe`. first abstraction should be done by os usually. so I would suggest to have `os: nd`.

rtolos replied:
As far as i know yes, `nd` is the official short name.
I ran some tests, as follows:
the device uses 2 types of connections - a `rest` connection to the `nd` software and an `ssh` connection straight to the linux machine.
With abstraction set to `os` and `os: nd` the `ssh` connection failed (the `rest` one worked properly).
With abstraction set to `platform`, `os: linux` and `platform: nd` both connections work.
I have limited knowledge on this platform / pyats entirely, but i will ask my more knowledgeable colleagues if it's possible to go with `os: nd`

tahigash replied:
I understand your point. currently, you need to have 2 devices for nd with different name like below in case you want to use 2 interfaces (CLI and REST)

```
devices:
  nd: # for CLI
    os: linux
  nd_test # for REST
    os: nd
```
if you really want to have only one device for both interfaces in testbed yaml, you might need to contribute Unicon.plugin for `nd`.

for example, `ise` plugin inherited `linux` plugin. something like this for `nd`.
https://github.com/CiscoTestAutomation/unicon.plugins/blob/master/src/unicon/plugins/ise/__init__.py

rtolos replied:
Thank you for the help, i changed ```os``` to ```nd``` and will work on adding a plugin in Unicon.plugin as well.

tahigash commented on src/rest/connector/libs/nd/implementation.py:
```diff
@@ -0,0 +1,310 @@
+        ------------
+            devices:
+                nd1:
+                    platform: nd
```
same

rtolos replied:
explained in a reply to the initial comment - platform is the correct abstraction selector for ND

tahigash replied:
same. `os: nd` is better.

tahigash replied:
replied on other comment thread.

rtolos replied:
Done

tahigash commented on src/rest/connector/libs/nd/implementation.py:
```diff
@@ -0,0 +1,310 @@
+        payload = {
+            "userName": username,
+            "userPasswd": password,
+            "domain": "DefaultAuth"  # CHANGE: domain
```
what `# CHANGE: domain` mean?

rtolos replied:
Sorry, some comments from the initial documentation / implementation phase slipped through.
Removed them.

tahigash commented on src/rest/connector/libs/nd/implementation.py:
```diff
@@ -0,0 +1,310 @@
+        """Decorator to make sure session to device is active
+
+           There is limitation on the amount of time the session cab be active
+           on the APIC. However, there are no way to verify if
```
not APIC...

rtolos replied:
Corrected, thank you

tahigash commented on src/rest/connector/tests/testbed.yaml:
```diff
@@ -51,6 +51,20 @@ devices:
           username: cisco
           password: cisco
+  nd:
+    os: linux
+    type: linux
+    platform: nd
+    custom:
+      abstraction:
+        order: [platform]
```
`os` should be `nd`.

tahigash replied:
same with above

rtolos replied:
Done

tahigash commented on src/rest/connector/libs/nd/implementation.py:
```diff
@@ -0,0 +1,310 @@
+
+
+class Implementation(Imp):
+    """Rest Implementation for ND
```
Please update doc as well.

rtolos replied:
Documentation updated

tahigash reviewed (changes requested):
re-reviewed and left comment about `os: nd`.

tahigash reviewed (commented):
replied to comments

rtolos commented:
Thank you for the review, I modified the files as requested.
In addition, I added retry logic to the commands (get / push / delete) - used mostly for timeout problems or unforseen circumstances - for example, if the device returns a message but the status code is not 200, there is no point in retrying, as our request reached but was most likely not valid.
Also added the logic to remove forward '/' from the ```api_url```, as it may cause problems and it was a very simple implementation step.

tahigash commented on docs/changelog/undistributed/changelog_nd_20210805110350.rst:
```diff
@@ -0,0 +1,3 @@
+
+* ND
```
```suggestion
* ND (Nexus Dashboard)
```

rtolos replied:
Changed, but with a twist:
It made more sense to me to have <long name> (<short name>) eg. Nexus Dashboard (ND)

tahigash commented on docs/user_guide/services/nd.rst:
```diff
@@ -0,0 +1,120 @@
+ND
+====
+
+The following services are supported by the REST connector for ND.
```
```suggestion
ND (Nexus Dashboard)
=================

The following services are supported by the REST connector for ND (Nexus Dashboard).
```

rtolos replied:
as above

tahigash reviewed (approved):
re-reviewed. just add new comment to have `Nexus Dashboard` for `nd` in changelog/doc. Please update, but approved in advance since it's small change.

thanks for your contribution.
